consul/pool.go

```diff
@@ -180,28 +187,81 @@ func (p *ConnPool) Shutdown() error {
 	return nil
 }
 
-// Acquire is used to get a connection that is
-// pooled or to return a new connection
+// acquire will return a pooled connection, if available. Otherwise it will
+// wait for an existing connection attempt to finish, if one if in progress,
+// and will return that one if it succeeds. If all else fails, it will return a
+// newly-created connection and add it to the pool.
 func (p *ConnPool) acquire(dc string, addr net.Addr, version int) (*Conn, error) {
-	// Check for a pooled ocnn
-	if conn := p.getPooled(addr, version); conn != nil {
-		return conn, nil
+	// markforUse does all the bookkeeping for an existing connection we wish
+	// to use from the pool.
+	markForUse := func(c *Conn) {
+		c.lastUsed = time.Now()
+		atomic.AddInt32(&c.refCount, 1)
 	}
 
-	// Create a new connection
-	return p.getNewConn(dc, addr, version)
-}
-
-// getPooled is used to return a pooled connection
-func (p *ConnPool) getPooled(addr net.Addr, version int) *Conn {
+	// Check to see if there's a pooled connection available. This is up
+	// here since it should the the vastly more common case than the rest
+	// of the code here.
 	p.Lock()
 	c := p.pool[addr.String()]
 	if c != nil {
-		c.lastUsed = time.Now()
-		atomic.AddInt32(&c.refCount, 1)
+		markForUse(c)
+		p.Unlock()
+		return c, nil
+	}
+
+	// If not (while we are still locked), set up the throttling structure
+	// for this address, which will make everyone else wait until our
+	// attempt is done.
+	var wait chan struct{}
+	var ok bool
+	if wait, ok = p.limiter[addr.String()]; !ok {
+		wait = make(chan struct{}, 1)
+		p.limiter[addr.String()] = wait
+	}
+	isLeadThread := !ok
+	p.Unlock()
+
+	// If we are the lead thread, make the new connection and then wake
+	// everybody else up to see if we got it.
+	if isLeadThread {
+		defer func() {
+			p.Lock()
+			delete(p.limiter, addr.String())
+			p.Unlock()
+
+			close(wait)
+		}()
+
+		c, err := p.getNewConn(dc, addr, version)
+		if err != nil {
```

Comment:
Super minor, but can we move the code in the defer to after `getNewConn` and delete the limiter entry and add the pool entry inside the same critical section? Just to save a defer and another round of locking